app/src/main/java/io/homeassistant/companion/android/controls/HaControlsProviderService.kt

```diff
@@ -77,41 +89,97 @@ class HaControlsProviderService : ControlsProviderService() {
+                        val entityRegistryFlow = webSocketRepository.getEntityRegistryUpdates()
                         // Load up initial values
                         // This should use the cached values that we should store in the DB.
                         // For now we'll use the rest API
+                        var areaRegistry = webSocketRepository.getAreaRegistry()
+                        var deviceRegistry = webSocketRepository.getDeviceRegistry()
+                        var entityRegistry = webSocketRepository.getEntityRegistry()
+                        val entities = mutableMapOf<String, Entity<Map<String, Any>>>()
                         controlIds.forEach {
                             val entity = integrationRepository.getEntity(it)
                             val domain = it.split(".")[0]
                             val control = domainToHaControl[domain]?.createControl(
                                 applicationContext,
-                                entity
+                                entity,
+                                getAreaForEntity(it, areaRegistry, deviceRegistry, entityRegistry)
                             )
+                            entities[it] = entity
                             subscriber.onNext(control)
                         }
 
                         // Listen for the state changed events.
-                        entityFlow.takeWhile { running }.collect {
-                            if (controlIds.contains(it.entityId)) {
-                                val domain = it.entityId.split(".")[0]
-                                val control = domainToHaControl[domain]?.createControl(
-                                    applicationContext,
-                                    it as Entity<Map<String, Any>>
-                                )
-                                subscriber.onNext(control)
+                        webSocketScope.launch {
+                            entityFlow.collect {
+                                if (controlIds.contains(it.entityId)) {
+                                    val domain = it.entityId.split(".")[0]
+                                    val control = domainToHaControl[domain]?.createControl(
+                                        applicationContext,
+                                        it as Entity<Map<String, Any>>,
+                                        getAreaForEntity(it.entityId, areaRegistry, deviceRegistry, entityRegistry)
+                                    )
+                                    subscriber.onNext(control)
+                                }
+                            }
+                        }
+                        webSocketScope.launch {
+                            areaRegistryFlow.collect {
+                                areaRegistry = webSocketRepository.getAreaRegistry()
+                                entities.forEach {
```

Comment:
There are 3 identical copies of this forEach loop.  I'd make that a small private function, and simply pass the different areaRegistry in to it each time, to make the code easier to follow.

Reply:
Good idea, I've also included the loop for loading the initial state as it was already almost identical.